src/Promitor.Agents.ResourceDiscovery/Scheduling/AzureResourceGroupsDiscoveryBackgroundJob.cs

```diff
@@ -30,11 +30,13 @@ public async Task ExecuteAsync(CancellationToken cancellationToken)
         {
             Logger.LogTrace("Discovering Azure Resource Groups...");
 
-            PagedPayload<AzureResourceGroupInformation> discoveredResourceGroups;
+            PagedPayload<AzureResourceGroupInformation> discoveredResourceGroups = null;
             do
             {
+                var currentPage = discoveredResourceGroups != null ? discoveredResourceGroups.PageInformation.CurrentPage + 1 : 0;
```

Comment:
There can be a nullref on `discoveredResourceGroups.PageInformation.CurrentPage`

We can also just do
```suggestion
            PagedPayload<AzureResourceGroupInformation> discoveredResourceGroups = null;
            do
            {
                var currentPage = 0;
                if(discoveredResourceGroups != null)
                {
                  if (discoveredResourceGroups.PageInformation == null)
                  {
                    throw new Exception("Current page information is not available and unable to determine next page.");
                  }
                  
                  currentPage = discoveredResourceGroups.PageInformation.CurrentPage + 1;
                }
```

However, we could also use a simple counter and increment it.

```suggestion
            PagedPayload<AzureResourceGroupInformation> discoveredResourceGroups = null;
            int currentPage = -1
            do
            {
                currentPage = currentPage + 1;
```

I'd use the latter though.

Reply:
The problem is a bit more nuanced and occurs even when the number of resources returned is less than 1000. This is due to the `HasMore` logic:
`public bool HasMore => (PageInformation.TotalRecords - (PageInformation.PageSize * PageInformation.CurrentPage)) > 0;`

So even if the total returned is less than 1000, `HasMore` will evaluate to true when starting with a `CurrentPage` value of `0`. Which leads me to a question - was it intentional to start with page 0 instead of page 1? I see indications in other areas of the code that seem to imply that page 1 should be the first value supplied for `currentPage`, e.g. in Scraper's `Discovery.ResourceDiscoveryRepository.GetResourceDiscoveryGroupAsync`, it starts with a `currentPage` value of `1`. If using a starting value of `1`, the `HasMore` logic evaluates as expected for returns less than the default `pageSize` of 1000.

Also note that the same issue exists in the `AzureSubscriptionDiscoveryBackgroundJob` as well.

Reply:
Additionally, this may not be in scope for this particular PR, but as part of investigating this I noticed that there is no deterministic sort/ordering applied to the Azure Resource Graph queries and am wondering whether that could lead to potential problems when the queries indeed return more than 1 page (i.e. 1000 records). 

Microsoft's [documentation](https://docs.microsoft.com/en-us/azure/governance/resource-graph/concepts/work-with-data) notes:
> When using Skip, it's recommended to order the results by at least one column with asc or desc. Without sorting, the results returned are random and not repeatable. If limit or take are used in the query, Skip is ignored.

Reply:
Both are valid points, let's handle the initial page to 1 in this PR and do the sorting in another one. Are you willing to jump on the sorting @gavinhc?